packages/core/src/utilities/VoxelManager.ts

```diff
@@ -167,46 +165,240 @@ export default class VoxelManager<T> {
   }
 
   /**
-   * Iterate over the points within the bounds, or the modified points if recorded.
+   * Iterates over the voxels in the VoxelManager and applies a callback function to each voxel.
+   * It can operate on IJK and LPS coordinate systems, and it can be limited to a specific region
+   * of the data if the isInObject function is provided.
+   *
+   * For the LPS calculations, both direction and spacing should be provided.
+   *
+   * If the boundsIJK is not provided, the iteration will be over the entire volume/data
+   *
+   *
+   * If the VoxelManager is backed by a Map, it will only iterate over the stored values.
+   * Otherwise, it will iterate over all voxels within the specified or default bounds.
    */
-  public forEach = (callback, options?) => {
+  public forEach = (
+    callback: (args: {
+      value: any;
+      index: number;
+      pointIJK: Point3;
+      pointLPS: Point3;
+    }) => void,
+    options?: {
+      boundsIJK?: BoundsIJK;
+      isInObject?: (pointLPS, pointIJK) => boolean;
+      returnPoints?: boolean;
+      imageData?: vtkImageData | CPUImageData;
+    }
+  ) => {
     const boundsIJK = options?.boundsIJK || this.getBoundsIJK();
-    const { isWithinObject } = options || {};
+    const isInObject = options.isInObject || this.isInObject || (() => true);
+    const returnPoints = options.returnPoints || true;
+
+    const useLPSTransform = options?.imageData;
+
+    const iMin = boundsIJK[0][0];
+    const iMax = boundsIJK[0][1];
+    const jMin = boundsIJK[1][0];
+    const jMax = boundsIJK[1][1];
+    const kMin = boundsIJK[2][0];
+    const kMax = boundsIJK[2][1];
+
+    const pointsInShape = [];
+
+    if (useLPSTransform) {
+      const { imageData } = options;
+      const direction = imageData.getDirection();
+      const rowCosines = direction.slice(0, 3);
+      const columnCosines = direction.slice(3, 6);
+      const scanAxisNormal = direction.slice(6, 9);
+
+      const spacing = imageData.getSpacing();
+      const [rowSpacing, columnSpacing, scanAxisSpacing] = spacing;
+
+      // @ts-ignore will
+
+      const start = vec3.fromValues(iMin, jMin, kMin);
+
+      // @ts-ignore will be fixed in vtk-master
+      const worldPosStart = imageData.indexToWorld(start);
+
+      const rowStep = vec3.fromValues(
+        rowCosines[0] * rowSpacing,
+        rowCosines[1] * rowSpacing,
+        rowCosines[2] * rowSpacing
+      );
+
+      const columnStep = vec3.fromValues(
+        columnCosines[0] * columnSpacing,
+        columnCosines[1] * columnSpacing,
+        columnCosines[2] * columnSpacing
+      );
+
+      const scanAxisStep = vec3.fromValues(
+        scanAxisNormal[0] * scanAxisSpacing,
+        scanAxisNormal[1] * scanAxisSpacing,
+        scanAxisNormal[2] * scanAxisSpacing
+      );
+
+      const currentPos = vec3.clone(worldPosStart) as Point3;
+
+      for (let k = kMin; k <= kMax; k++) {
+        const startPosJ = vec3.clone(currentPos);
```

Comment:
There is an optimized version of this in the currently closed source code - it is quite noticeably faster as it pre-generates i,j,k arrays and adds the three numbers for each dimension.  That version actually works with sparse data voxel managers, whereas this one doesn't.  Maybe it is worth considering re-implementing that?  Lets leave it for now, but think about doing it later as it is really just a performance optimization, but the client for that closed source might be willing to contribute just that code back so it is easy for them to update in the future.

Reply:
Added a comment